src/mlpack/core/tree/cellbound.hpp

```diff
@@ -0,0 +1,306 @@
+   *
+   * @param point Point to which the minimum distance is requested.
+   */
+  template<typename VecType>
+  ElemType MinDistance(const VecType& point,
+                       typename boost::enable_if<IsVector<VecType>>* = 0) const;
+
+  /**
+   * Calculates minimum bound-to-bound distance.
+   *
+   * @param other Bound to which the minimum distance is requested.
+   */
+  ElemType MinDistance(const CellBound& other) const;
+
+  /**
+   * Calculates maximum bound-to-point squared distance.
+   *
+   * @param point Point to which the maximum distance is requested.
+   */
+  template<typename VecType>
+  ElemType MaxDistance(const VecType& point,
+                       typename boost::enable_if<IsVector<VecType>>* = 0) const;
+
+  /**
+   * Computes maximum distance.
+   *
+   * @param other Bound to which the maximum distance is requested.
+   */
+  ElemType MaxDistance(const CellBound& other) const;
+
+  /**
+   * Calculates minimum and maximum bound-to-bound distance.
+   *
+   * @param other Bound to which the minimum and maximum distances are
+   *     requested.
+   */
+  math::RangeType<ElemType> RangeDistance(const CellBound& other) const;
+
+  /**
+   * Calculates minimum and maximum bound-to-point distance.
+   *
+   * @param point Point to which the minimum and maximum distances are
+   *     requested.
+   */
+  template<typename VecType>
+  math::RangeType<ElemType> RangeDistance(
+      const VecType& point,
+      typename boost::enable_if<IsVector<VecType>>* = 0) const;
+
+  /**
+   * Expands this region to include new points.
+   *
+   * @tparam MatType Type of matrix; could be Mat, SpMat, a subview, or just a
+   *   vector.
+   * @param data Data points to expand this region to include.
+   */
+  template<typename MatType>
+  CellBound& operator|=(const MatType& data);
+
+  /**
+   * Expands this region to encompass another bound.
+   */
+  CellBound& operator|=(const CellBound& other);
+
+  /**
+   * Determines if a point is within this bound.
+   */
+  template<typename VecType>
+  bool Contains(const VecType& point) const;
+
+  /**
+   * Calculate the bounds of all subrectangles. You should set the lower and the
+   * high addresses.
+   *
+   * @param data Points that are contained in the node.
+   */
+  template<typename MatType>
+  void UpdateAddressBounds(const MatType& data);
+
+  /**
+   * Returns the diameter of the hyperrectangle (that is, the longest diagonal).
+   */
+  ElemType Diameter() const;
+
+  /**
+   * Serialize the bound object.
+   */
+  template<typename Archive>
+  void Serialize(Archive& ar, const unsigned int version);
+
+ private:
+  //! The precision of the tree element type.
+  static constexpr size_t order = sizeof(AddressElemType) * CHAR_BIT;
+  //! Maximum number of subrectangles.
+  const size_t maxNumBounds = 10;
+  //! The dimensionality of the bound.
+  size_t dim;
+  //! The bounds for each dimension.
+  math::RangeType<ElemType>* bounds;
```

Comment:
It seems like we can't remove this because of the `Center()` calculation and the `Diameter()`.  Removing it would mean that for `Center()` we would have to calculate the mean of all subrectangle centers, and it would be harder to cache something for `Diameter()`.  My idea in removing it was to make `sizeof(CellBound)` smaller but it seems that's not really feasible.  But I thought it would be useful to write down the notes anyway in case it was ever revisited later. :)